File: Data_Structure/LinkedList_x1/deleteNode_LinkedList.cpp
#include<iostream>
using namespace std;

class Node
{
    public:
    int data;
    Node *next;
};

class LinkedList
{
    Node *head;

    public:

    LinkedList()
    {
        head = NULL;
    }

    void addLast(int val)
    {
        Node *newNode = new Node();
        newNode->data = val;
        newNode->next = NULL;

        if(head == NULL)
            head = newNode;
        else
        {
            Node *lastNode = head;

            while(lastNode->next != NULL)
            {
                lastNode = lastNode->next;
            }

            lastNode->next = newNode;
        }
    }

    void del(int key)
    {
        Node *temp;

        if(head->data == key)
        {
            temp = head;
            head = head->next;
            delete temp;
        }
        else
        {
            Node *current = head;

            while(current->next != NULL)
            {
                if(current->next->data == key)
                {
                    temp = current->next;
                    current->next = current->next->next;
                    delete temp;
                    break;
                }
                else
                    current = current->next;
            }
        }
    }

    void print()
    {
        Node *temp = head;

        while(temp != NULL)
        {
            cout<<temp->data<< "=>\t";
            temp = temp->next;
        }
        cout << "NULL\n";
    }
};

int main()
{
    LinkedList list;

    list.addLast(10);
    list.addLast(20);
    list.addLast(30);
    list.addLast(40);

    cout<<"The LinkedList Elements Are [Before Deletion]:"<<endl;
    list.print();

    cout<<"Deleting Element: "<<30<<endl;
    list.del(30);

    cout<<"The LinkedList Elements Are [After Deletion]:"<<endl;
    list.print();

    cout << endl;

    return 0;
}

File: JS_Algorithms/DeleteNodelinkedList.js
/* Delete @Node in a Singly LinkedList */

var head = null;

class Node
{
    constructor(val)
     {
            this.data = val;
            this.next = null;
      }
}

function print()
{
    var temp = head;
    while(temp != null)
    {
        document.write(temp.data+"<br>");
        temp = temp.next;
    }
}

function addLast(val)
{
    var newNode = new Node(val);

    if(head == null)
        head = newNode;
    else
    {
        var lastNode = head;

        while(lastNode.next != null)
        {
            lastNode = lastNode.next;
        }

        lastNode.next = newNode;
    }
}

function del(key)
{
    if(head.data == key)
    {
        head = head.next;
    }
    else
    {
        var temp = head;

        while(temp.next != null)
        {
            if(temp.next.data == key)
            {
                temp.next = temp.next.next;
                break;
            }
            else
                temp = temp.next;
        }
    }
}

addLast(10);
addLast(20);
addLast(30);
addLast(40);


document.write("The LinkedList Elements Are [Before Deletion]:<br>");
print();

document.write("Deleting Element: "+30+"<br>");
del(30);

document.write("The LinkedList Elements Are [After Deletion]:<br>");

print();




